Reset the socket after a timeout to make sure no wrong data is received

marcdejonge opened the pull request:
Fixes #1222 
Fixes #1418 

I did the following stuff on a timeout:

- Force disconnect the current socket, so even a persistent connection is cleared (so no chance for illegal data to be returned the next time).
- Change the lazy_connect logic to reset this correctly. Without this change, the loop in `cluster_send_command` will run very quickly without really retrying to connect.

I'm not really sure if this last change is correct in all cases, but it seems to work for me.

yatsukhnenko commented:
@marcdejonge thank you for PR. I'm ok with disconnecting socket after timeout but I'm not sure about changing lazy_connect value

marcdejonge commented:
Ok, what is the use for the lazy_connect? If this change isn't made, the disconnected socket will never reconnect automatically. For what other things is it used?

marcdejonge commented:
I've also tried to reconnect immediately after the disconnect, but when the interruption is longer than 1 timeout then it will never reconnect correctly.

yatsukhnenko commented:
In my mind `lazy_connect` is only needed to not connect while creating `redis_sock` object and it should not be used elsewhere.
@michael-grunder what if we will change `redis_sock_get` function to check `redis_sock->status` instead of `redis_sock->lazy_connect`?

marcdejonge commented:
I've been digging into this lazy_connect variable and it only seems to be really used the cluster implementation where the socket is not opened immediately to all the nodes. But it seems to me that not using this variable and using the `redis_sock->status` works more reliable when trying to reconnect to an existing socket. The you can just call the `redis_sock_server_open` method every time, since that method already checks the status.

The other solution would be to remove the socket completely when disconnecting to create a new socket. But I'm not sure how to do that exactly.

yatsukhnenko commented:
Since @michael-grunder has a secret knowledge :smile: in this project I want to wait and listen his thoughts about this. Definitely we can't just remove `lazy_connect` from `redis_sock` structure.

michael-grunder commented:
Hey everyone, just catching up on the thread.

What if we localize `redis_sock->lazy_connect = 1` to after we manually disconnect because of a timeout?  That way it doesn't risk changing behavior any time we call `redis_sock_disconnect` at all.

marcdejonge commented:
I tried that, but it won't work in all cases. For example, if the next
command tries to reconnect and it fails, the lazy_connect will stay on 0
and it won't try again. Then we are in a state where it never connect again
until you restart the php price. This was the whole reason why I changed
that piece of code.

On Thu, Oct 4, 2018, 19:51 Michael Grunder <notifications@github.com> wrote:

> Hey everyone, just catching up on the thread.
>
> What if we localize redis_sock->lazy_connect = 1 to after we manually
> disconnect because of a timeout? That way it doesn't risk changing behavior
> any time we call redis_sock_disconnect at all.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/phpredis/phpredis/pull/1417#issuecomment-427110693>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAVqVSa1tsxduT4H6HtdTA39GI8F2wyDks5uhkqvgaJpZM4XFpSp>
> .
>

yatsukhnenko commented:
> Definitely we can't just remove lazy_connect from redis_sock structure.

Hmm, I was wrong. It seems that we can remove it. I need to check everything again :)

marcdejonge commented:
Here you go, removed the lazy_connect completely. I think this still works correctly.

yatsukhnenko commented:
Thanks, @marcdejonge! I think we should check the result of call `redis_sock_server_open`

yatsukhnenko commented:
And `c->cmd_sock->stream` check is unnecessary`

marcdejonge commented:
I've fixed these issues. The `redis_sock_server_open` call I've added to the CLUSTER_SEND_PAYLOAD since this is almost always called directly after each other. All tests seem ok here.

yatsukhnenko commented:
@marcdejonge I received email notification but cant find the comment from Jonathan Champ here. He noticed that you don't path TSRMLS_CC to all calls of `redis_sock_server_open`. Fix please.

yatsukhnenko commented:
@marcdejonge sorry, that comment is not relevant anymore :)

yatsukhnenko commented:
@michael-grunder any objections to merge this change?

yatsukhnenko commented:
@michael-grunder ping

liamsorsby commented:
As per the comment I've just made on issue page #1222 . I'm still able to replicate the issue.
